rust/agama-logs/src/main.rs

```diff
@@ -0,0 +1,221 @@
+
+    fn to(&self) -> PathBuf {
+        // remove leading '/' if any from the path (reason see later)
+        let r_path = Path::new(self.src_path.as_str()).strip_prefix("/").unwrap();
+
+        // here is the reason, join overwrites the content if the joined path is absolute
+        self.dst_path.join(r_path)
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        // for now keep directory structure close to the original
+        // e.g. what was in /etc will be in /<tmp dir>/etc/
+        fs::create_dir_all(self.to().parent().unwrap())?;
+
+        let options = CopyOptions::new();
+        // fs_extra's own Error doesn't implement From trait so ? operator is unusable
+        match copy_items(
+            &[self.src_path.as_str()],
+            self.to().parent().unwrap(),
+            &options,
+        ) {
+            Ok(_p) => Ok(()),
+            Err(_e) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Copying of a file failed",
+            )),
+        }
+    }
+}
+
+impl LogItem for LogCmd {
+    fn from(&self) -> &String {
+        &self.cmd
+    }
+
+    fn to(&self) -> PathBuf {
+        self.dst_path.as_path().join(format!("{}", self.cmd))
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        let cmd_parts = self.cmd.split_whitespace().collect::<Vec<&str>>();
+        let file_path = self.to();
+        let output = Command::new(cmd_parts[0])
+            .args(cmd_parts[1..].iter())
+            .output()
+            .expect("Failed to run the command");
+        let mut file_stdout = File::create(format!("{}.out.log", file_path.display()))?;
+        let mut file_stderr = File::create(format!("{}.err.log", file_path.display()))?;
+
+        file_stdout.write_all(&output.stdout)?;
+        file_stderr.write_all(&output.stderr)?;
+
+        Ok(())
+    }
+}
+
+fn main() -> Result<(), io::Error> {
+    if !Uid::effective().is_root() {
+        panic!("No Root, no logs. Sorry.");
+    }
+
+    // 0. preparation, e.g. in later features some log commands can be added / excluded per users request or
+    let commands = DEFAULT_COMMANDS;
+    let paths = DEFAULT_PATHS;
+    let result = format!("{}.{}", DEFAULT_RESULT, DEFAULT_COMPRESSION.1);
+    let noisy = DEFAULT_NOISY;
+    let compression = DEFAULT_COMPRESSION.0;
+    let mut log_sources: Vec<Box<dyn LogItem>> = Vec::new();
+
+    showln!(noisy, "Collecting Agama logs:");
+
+    // 1. create temporary directory where to collect all files (similar to what old save_y2logs
+    // does)
+    let tmp_dir = TempDir::new(DEFAULT_TMP_DIR)?;
+
+    // 2. collect existing / requested paths which should already exist
+    showln!(noisy, "\t- proceeding well known paths");
+    for path in paths {
+        // assumption: path is full path
+        if Path::new(path).try_exists().is_ok() {
+            log_sources.push(Box::new(LogPath {
+                src_path: path.to_string(),
+                dst_path: tmp_dir.path().to_path_buf(),
```

Comment:
You can replace `to_path_buf()` with `to_owned()` which express your intent better. If you prefer, you can have a constructor that takes care of the conversion details:

```
impl LogPath {
  fn new(src: &str, dst: &Path) -> Self { 
    Self { src_path: src.to_string(), dst_path: dst.to_owned() }
  }
}
```

And you can use it as:

```rust
  log_sources.push(Box::new(LogPath::new(path, tmp_dir.path()))
```

Reply:
nice ... some more reading to the list ;-)